File: Daily/sbgo/Leetcode1074-元素和为目标值的子矩阵数量.go
package sbgo

func numSubmatrixSumTarget(matrix [][]int, target int) int {
	ans, n, m := 0, len(matrix), len(matrix[0])
	count := func(presum []int, target int) int {
		res, cur := 0, 0
		mp := map[int]int{}
		mp[0] = 1
		for i := 0; i < len(presum); i++ {
			cur += presum[i]
			cnt, ok := mp[cur-target]
			if ok {
				res += cnt
			}
			mp[cur]++
		}
		return res
	}
	for i := 0; i < n; i++ {
		presum := make([]int, m)
		for j := i; j < n; j++ {
			for k := 0; k < m; k++ {
				presum[k] += matrix[j][k]
			}
			ans += count(presum, target)
		}
	}
	return ans
}

File: Leetcode/Rust/Leetcode1074-元素和为目标值的子矩阵数量.rs
use std::collections::HashMap;
impl Solution {
    pub fn num_submatrix_sum_target(matrix: Vec<Vec<i32>>, target: i32) -> i32 {
        let (n,m,mut ans) = (matrix.len(),matrix[0].len(),0);
        for i in 0..n{
            let mut presum = vec![0;m];
            for j in i..n{
                for k in 0..m{
                    presum[k] += matrix[j][k];
                }
                let mut mp = HashMap::new();
                mp.insert(0, 1);
                let (mut res,mut cur) = (0,0);
                for idx in 0..m{
                    cur += presum[idx];
                    if mp.contains_key(&(cur - target)){
                        res += mp[&(cur -target)];
                    }
                    let h = mp.entry(cur).or_insert(0);
                    *h += 1;
                }
                ans += res;
            }
        }
        ans
    }
}